llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -433,6 +433,19 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
       indicatePessimisticFixpoint();
       return;
     }
+
+    bool HasAllocaOrASCast = false;
+    for (BasicBlock &BB : *F) {
+      for (Instruction &I : BB) {
+        if (isa<AllocaInst>(I) || isa<AddrSpaceCastInst>(I)) {
+          HasAllocaOrASCast = true;
+          removeAssumedBits(FLAT_SCRATCH_INIT);
+          break;
+        }
+      }
+      if (HasAllocaOrASCast)
+        break;
+    }
```

Comment:
At some point, you need to look at the non-call instructions.
It makes sense to do that in update *if* we ever do value simplification or dead code elimination as part of the pass. Otherwise, init is fine, I think.